usr/opt/OSC200/usr/share/man/man3/rpc.3
.\" This manpage source uses rsml coding. 
.so /usr/share/lib/tmac/sml
.so /usr/share/lib/tmac/rsml
.\"
.\"
.\" @DEC_COPYRIGHT@
.\"
.\"
.\" HISTORY
.\" $Log: rpc.3,v $
.\" Revision 3.2.4.3  1993/05/26  15:20:36  Leslie_Wharton
.\" 	Fixed fatal tman errors
.\" 	[1993/05/26  15:19:46  Leslie_Wharton]
.\"
.\" Revision 3.2.4.2  1993/05/26  14:19:44  Joyce_Demers
.\" 	Connected line in the NAME section so that all the appropriate pointer files will be created.  Also changed .5 parameters to .sp macros to be 0.5.
.\" 	[1993/05/26  14:17:56  Joyce_Demers]
.\" 
.\" Revision 3.2.2.4  1993/02/04  13:17:53  Ken_Hall
.\" 	format fixes
.\" 	[93/02/04  11:25:48  Ken_Hall]
.\" 
.\" 	Retrofitted changes since IFT
.\" 	[92/07/22  11:19:59  Joyce_Demers]
.\" 
.\" Revision 3.2.2.3  92/07/22  12:25:03  Joyce_Demers
.\" 	"Retrofitted changes since IFT"
.\" 
.\" Revision 3.2.2.2  92/05/27  11:15:57  Joyce_Demers
.\" 	"Changes from port_dec_refpages script"
.\" 
.\" Revision 3.2.1.2  92/05/22  13:24:10  Joyce_Demers
.\" 	Changes from port_dec_refpage script
.\" 
.\" Revision 3.2  92/04/30  12:20:49  devrcs
.\" 	Initial load of project
.\" 
.\" $EndLog$
.\"
.\" @(#)$RCSfile: rpc.3,v $ $Revision: 3.2.4.3 $ (DEC) $Date: 1993/05/26 15:20:36 $
.\" 
.\" @(#)rpc.3n	1.1 88/06/02 4.0NFSSRC SMI; from UCB X.X
.\" @(#) from SUN 1.14
.TH rpc 3
.SH NAME
.PP
rpc, auth_destroy, authnon_create, authunix_create, authunix_create_default, callrpc,clnt_broadcast, clnt_call, clnt_create, clnt_control, clnt_destroy, clnt_freeres, clnt_geterr, clnt_pcreateerror, clnt_perrno, clnt_perror, clnt_spcreateerror, clnt_sperrno, clnt_sperror, clntraw_create, clnttcp_create, clntudp_create, get_myaddress, getnetname, host2netname, netname2host, netname2user, pmap_getmaps, pmap_getport, pmap_rmtcall, pmap_set, pmap_unset, registerrpc, rpc_createrr, svc_destroy, svc_fdset, svc_freeargs, svc_getargs, svc_getcaller, svc_getreq, svc_getreqset, svc_register, svc_run, svc_sendreply, svc_unregister, svcerr_auth, svcerr_decode, svcerr_noproc, svcerr_noprog, svcerr_progvers, svcerr_systemerr, svcerr_weakauth, svcfd_create, svcraw_create, svctcp_create, svcudp_create, usr2netname, xdr_accepted_reply, xdr_authunix_parms, xdr_callhdr, xdr_callmsg, xdr_opaque_auth, xdr_pmap, xdr_pmaplist, xdr_rejected_reply, xdr_replymsg, xprt_register, xprt_unregister, rpc_functions \- library routines for ONC remote procedure calls
.SH SYNOPSIS
.PP
.ft B
.nf
.sp 0.5
#include <rpc/rpc.h>
.fi
.ft R
.br
.if t .ne 8
.PP
.ft B
.nf
.sp 0.5
void
auth_destroy(auth)
\s-1AUTH\s0 *auth;
.fi
.ft R
.IP
A macro that destroys the authentication information associated with
.IR auth .
Destruction usually involves deallocation of private data
structures. The use of
\*Vauth\*O
is undefined after calling
.BR auth_destroy(\|) .
.br
.if t .ne 6
.PP
.ft B
.nf
.sp 0.5
\s-1AUTH\s0 *
authnone_create(\|)
.fi
.ft R
.IP
Create and returns an
.SM RPC
authentication handle that passes nonusable authentication
information with each remote procedure call. This is the
default authentication used by
.SM ONC RPC.
.if t .ne 22
.PP
.ft B
.nf
.\".sp 0.5
.\"\s-1AUTH\s0 *
.\"authdes_create(name, window, syncaddr, ckey)
.\"char *name;
.\"unsigned window;
.\"struct sockaddr *addr;
.\"des_block *ckey;
.\".fi
.\".ft R
.\".IP
.\".B authdes_create(\|)
.\"is the first of two routines which interface to the
.\".SM RPC
.\"secure authentication system, known as
.\".SM DES
.\"authentication.
.\"The second is
.\".BR authdes_getucred(\|) ,
.\"below. Note: the keyserver daemon
.\".BR keyserv (8)
.\"must be running for the
.\".SM DES
.\"authentication system to work.
.\".IP
.\".BR authdes_create(\|) ,
.\"used on the client side, returns an authentication handle that
.\"will enable the use of the secure authentication system.
.\"The first parameter
.\".I name
.\"is the network name, or
.\".IR netname ,
.\"of the owner of the server process. This field usually
.\"represents a
.\".I hostname
.\"derived from the utility routine
.\".BR host2netname ,
.\"but could also represent a user name using
.\".BR user2netname .
.\"The second field is window on the validity of
.\"the client credential, given in seconds.  A small
.\"window is more secure than a large one, but choosing
.\"too small of a window will increase the frequency of
.\"resynchronizations because of clock drift. The third
.\"parameter
.\".I syncaddr
.\"is optional.  If it is
.\".SM NULL\s0,
.\"then the authentication system will assume
.\"that the local clock is always in sync with the server's
.\"clock, and will not attempt resynchronizations. If an address
.\"is supplied, however, then the system will use the address
.\"for consulting the remote time service whenever
.\"resynchronization
.\"is required. This parameter is usually the
.\"address of the
.\".SM RPC
.\"server itself. The final parameter
.\".I ckey
.\"is also optional.  If it is
.\".SM NULL\s0,
.\"then the authentication system will
.\"generate a random
.\".SM DES
.\"key to be used for the encryption of credentials.
.\"If it is supplied, however, then it will be used instead.
.\".br
.\".if t .ne 13
.\".PP
.\".ft B
.\".nf
.\".sp 0.5
.\"authdes_getucred(adc, uid, gid, grouplen, groups)
.\"struct authdes_cred *adc;
.\"short *uid;
.\"short *gid;
.\"short *grouplen;
.\"int *groups;
.\".fi
.\".ft R
.\".IP
.\".BR authdes_getucred(\|) ,
.\"the second of the two
.\".SM DES
.\"authentication routines,
.\"is used on the server side for converting a
.\".SM DES
.\"credential, which is
.\"operating system independent, into a
.\".UX
.\"credential. This routine differs from utility routine
.\".B netname2user
.\"in that
.\".B authdes_getucred(\|)
.\"pulls its information from a cache, and does not have to do a
.\"Yellow Pages lookup everytime it is called to get its information.
.\".br
.\".if t .ne 10
.\".PP
.\".ft B
.\".nf
.sp 0.5
\s-1AUTH\s0 *
authunix_create(host, uid, gid, len, aup_gids)
char *host;
int uid, gid, len, *aup.gids;
.fi
.ft R
.IP
Create and return an
.SM ONC RPC
authentication handle that contains
.\".UX
authentication information.
The parameter
\*Vhost\*O
is the name of the machine on which the information was
created;
\*Vuid\*O
is the user's user
.SM ID ;
\*Vgid\*O
is the user's current group
.SM ID ;
\*Vlen\*O
and
\*Vaup_gids\*O
refer to a counted array of groups to which the user belongs.
It is easy to impersonate a user.
.br
.if t .ne 5
.PP
.ft B
.nf
.sp 0.5
\s-1AUTH\s0 *
authunix_create_default(\|)
.fi
.ft R
.IP
Calls
.B authunix_create(\|)
with the appropriate parameters.
.br
.if t .ne 13
.PP
.ft B
.nf
.sp 0.5
callrpc(host, prognum, versnum, procnum, inproc, in, outproc, out)
char *host;
u_long prognum, versnum, procnum;
char *in, *out;
xdrproc_t inproc, outproc;
.fi
.ft R
.IP
Call the remote procedure associated with
.IR prognum ,
.IR versnum ,
and
\*Vprocnum\*O
on the machine,
.IR host .
The parameter
\*Vin\*O
is the address of the procedure's argument(s), and
\*Vout\*O
is the address of where to place the result(s);
\*Vinproc\*O
is used to encode the procedure's parameters, and
\*Voutproc\*O
is used to decode the procedure's results.
This routine returns zero if it succeeds, or the value of
.B "enum clnt_stat"
cast to an integer if it fails.
The routine
.B clnt_perrno(\|)
is handy for translating failure statuses into messages.
.IP
Warning: calling remote procedures with this routine
uses
.SM UDP/IP
as a transport; see
.B clntudp_create(\|)
for restrictions.
You do not have control of timeouts or authentication using
this routine.
.br
.if t .ne 16
.PP
.ft B
.nf
.sp 0.5
enum clnt_stat
clnt_broadcast(prognum, versnum, procnum, inproc, in, outproc, out,
.br
   eachresult)
u_long prognum, versnum, procnum;
char *in, *out;
xdrproc_t inproc, outproc;
resultproc_t eachresult;
.fi
.ft R
.IP
Like
.BR callrpc(\|) ,
except the call message is broadcast to all locally
connected broadcast nets. Each time it receives a
response, this routine calls
.BR eachresult(\|) ,
whose form is:
.IP
.RS 1i
.ft B
.nf
eachresult(out, addr)
char *out;
struct sockaddr_in *addr;
.ft R
.fi
.RE
.IP
where
\*Vout\*O
is the same as
\*Vout\*O
passed to
.BR clnt_broadcast(\|) ,
except that the remote procedure's output is decoded there;
\*Vaddr\*O
points to the address of the machine that sent the results.
If
.B eachresult(\|)
returns zero,
.B clnt_broadcast(\|)
waits for more replies; otherwise it returns with appropriate
status. If
.B eachresult(\|)
is
.SM NULL
.B clnt_broadcast(\|)
returns without waiting for any replies.
.IP
Warning: broadcast sockets are limited in size to the
maximum transfer unit of the data link. For Ethernet,
the caller's argument size should not exceed 1400 bytes.
.br
.if t .ne 13
.PP
.ft B
.nf
.sp 0.5
enum clnt_stat
clnt_call(clnt, procnum, inproc, in, outproc, out, tout)
\s-1CLIENT\s0 *clnt;
u_long
procnum;
xdrproc_t inproc, outproc;
char *in, *out;
struct timeval tout;
.fi
.ft R
.IP
A macro that calls the remote procedure
\*Vprocnum\*O
associated with the client handle,
.IR clnt ,
which is obtained with an
.SM RPC
client creation routine such as
.BR clnt_create(\|) .
The parameter
\*Vin\*O
is the address of the procedure's argument(s), and
\*Vout\*O
is the address of where to place the result(s);
\*Vinproc\*O
is used to encode the procedure's parameters, and
\*Voutproc\*O
is used to decode the procedure's results;
\*Vtout\*O
is the time allowed for results to come back.
.br
.if t .ne 7
.PP
.ft B
.nf
.sp 0.5
clnt_destroy(clnt)
\s-1CLIENT\s0 *clnt;
.fi
.ft R
.IP
A macro that destroys the client's
.SM RPC
handle. Destruction usually involves deallocation
of private data structures, including
\*Vclnt\*O
itself.  Use of
\*Vclnt\*O
is undefined after calling
.BR clnt_destroy(\|) .
If the
.SM RPC
library opened the associated socket, it will close it also.
Otherwise, the socket remains open.
.br
.if t .ne 10
.PP
.ft B
.nf
.sp 0.5
\s-1CLIENT\s0 *
clnt_create(host, prog, vers, proto)
char *host;
u_long prog, vers;
char *proto;
.fi
.ft R
.IP
Generic client creation routine.
\*Vhost\*O
identifies the name of the remote host where the server
is located.
\*Vproto\*O
indicates which kind of transport protocol to use. The
currently supported values for this field are \(lqudp\(rq
and \(lqtcp\(rq.
Default timeouts are set, but can be modified using
.BR clnt_control(\|) .
.IP
Warning: Using
.SM UDP
has its shortcomings.  Since
.SM UDP\s0-based
.SM RPC
messages can only hold up to 8 Kbytes of encoded data,
this transport cannot be used for procedures that take
large arguments or return huge results.
.br
.if t .ne 10
.PP
.ft B
.nf
.sp 0.5
bool_t
clnt_control(cl, req, info)
\s-1CLIENT\s0 *cl;
int request;
char *info;
.fi
.ft R
.IP
A macro used to change or retrieve various information
about a client object.
\*Vreq\*O
indicates the type of operation, and
\*Vinfo\*O
is a pointer to the information. For both
.SM UDP
and
.SM TCP\s0,
the supported values of
\*Vreq\*O
and their argument types and what they do are:
.IP
.TS
tab(~);
l l l.
CLSET_TIMEOUT~struct timeval~set total timeout
CLGET_TIMEOUT~struct timeval~get total timeout
CLGET_FD~int~get associated socket
CLSET_FD_CLOSE~void~T{
close socket on \f3clnt_destroy(\|)\f1
T}
CLSET_FD_NOCLOSE~void~T{

usr/opt/OSC200/src/usr/sbin/rstatd/rpc.rstatd.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
#ifndef lint
static char *rcsid = "@(#)$RCSfile: rpc.rstatd.c,v $ $Revision: 1.1.2.2 $ (DEC) $Date: 1993/05/26 17:47:38 $";
#endif

/*
 * Copyright (c) 1984 by Sun Microsystems, Inc.
 */

/* 
 * rstat demon: called from inetd, or self-started
 */

#include <signal.h>
#include <stdio.h>
#include <rpc/rpc.h>
#include <sys/socket.h>
#include <nlist.h>
#include <sys/errno.h>
#include "vmmeter.h"
#include <net/if.h>
#include <sys/time.h>
#include <rpcsvc/rstat.h>
#include <syslog.h>

#define CLOSEDOWN 20		/* how long to wait before do_exiting */
#ifndef FSCALE
#define FSCALE (1 << 8)
#endif

struct nlist nl[] = {
#define	X_CPTIME	0
	{ "_cp_time" },
#define	X_SUM		1
	{ "_sum" },
#define	X_IFNET		2
	{ "_ifnet" },
#define	X_DKXFER	3
	{ "_dk_xfer" },
#define	X_BOOTTIME	4
	{ "_boottime" },
#define	X_AVENRUN	5
	{ "_avenrun" },
#define X_HZ		6
	{ "_hz" },
	"",
};

static void rstatprog_2();
static void rstatprog_3();
static void rstatprog_4();
int updatestat();
int stats_service();
void setup();

static int _rpcpmstart;		/* Started by a port monitor ? */
static int _rpcfdtype;		/* Whether Stream or Datagram ? */
int kmem;
unsigned long firstifnet, numintfs;	/* chain of ethernet interfaces */
int sincelastreq = 0;		/* number of alarms since last request */
extern int errno;
#ifdef vax
double avrun[3];
#endif /*vax*/

main()
{
	register SVCXPRT *transp;
	int sock;
	int proto;
	struct sockaddr_in saddr;
	int asize = sizeof(saddr);

	if (getsockname(0, (struct sockaddr *)&saddr, &asize) == 0) {
		int ssize = sizeof(int);

		if (saddr.sin_family != AF_INET)
			exit(1);

		if (getsockopt(0, SOL_SOCKET, SO_TYPE,
				(char *)&_rpcfdtype, &ssize) == -1) {
		  printf ( "getsocketopt error \n");
		  exit(1);
		}

		sock = 0;
		_rpcpmstart = 1;
		proto = 0;
	} else {
#ifndef RPC_SVC_FG
		int pid;

		pid = fork();
		if (pid < 0) {
			perror("cannot fork");
			exit(1);
		}
		if (pid)
			exit(0);
#endif
		sock = RPC_ANYSOCK;
		(void) pmap_unset(RSTATPROG, RSTATVERS_SWTCH);
		(void) pmap_unset(RSTATPROG, RSTATVERS_TIME);
		(void) pmap_unset(RSTATPROG, RSTATVERS_VAR);
	}

	if ((_rpcfdtype == 0) || (_rpcfdtype == SOCK_DGRAM)) {
		transp = svcudp_create(sock);
		if (transp == NULL) {
			_msgout("cannot create udp service.");
			exit(1);
		}
		if (!_rpcpmstart)
			proto = IPPROTO_UDP;
		if (!svc_register(transp, RSTATPROG, RSTATVERS_SWTCH, rstatprog_2, proto)) {
			_msgout("unable to register (RSTATPROG, RSTATVERS_SWTCH, udp).");
			exit(1);
		}
		if (!svc_register(transp, RSTATPROG, RSTATVERS_TIME, rstatprog_3, proto)) {
			_msgout("unable to register (RSTATPROG, RSTATVERS_TIME, udp).");
			exit(1);
		}
		if (!svc_register(transp, RSTATPROG, RSTATVERS_VAR, rstatprog_4, proto)) {
			_msgout("unable to register (RSTATPROG, RSTATVERS_VAR, udp).");
			exit(1);
		}
	}

	if (transp == (SVCXPRT *)NULL) {
		_msgout("could not create a handle");
		exit(1);
	}
	/*
	 * This portion of the code is not generated by rpcgen and is added.
	 */
	setup();
	updatestat();
	alarm(1);
	signal(SIGALRM, updatestat);
	svc_run();
	_msgout("svc_run returned");
	exit(1);
	/* NOTREACHED */
}
static void
rstatprog_2(rqstp, transp)
	struct svc_req *rqstp;
	register SVCXPRT *transp;
{
	union {
		int fill;
	} argument;
	char *result;
	bool_t (*xdr_argument)(), (*xdr_result)();
	char *(*local)();

	switch (rqstp->rq_proc) {
	case NULLPROC:
		(void) svc_sendreply(transp, xdr_void, (char *)NULL);
		return;

	case RSTATPROC_STATS:
		xdr_argument = xdr_void;
		xdr_result = xdr_statsswtch;
		local = (char *(*)()) rstatproc_stats_2;
		break;

	case RSTATPROC_HAVEDISK:
		xdr_argument = xdr_void;
		xdr_result = xdr_long;
		local = (char *(*)()) rstatproc_havedisk_2;
		break;

	default:
		svcerr_noproc(transp);
		return;
	}
	bzero((char *)&argument, sizeof(argument));
	if (!svc_getargs(transp, xdr_argument, &argument)) {
		svcerr_decode(transp);
		return;
	}
	result = (*local)(&argument, rqstp);
	if (result != NULL && !svc_sendreply(transp, xdr_result, result)) {
		svcerr_systemerr(transp);
	}
	if (!svc_freeargs(transp, xdr_argument, &argument)) {
		_msgout("unable to free arguments");
		exit(1);
	}
	return;
}

static void
rstatprog_3(rqstp, transp)
	struct svc_req *rqstp;
	register SVCXPRT *transp;
{
	union {
		int fill;
	} argument;
	char *result;
	bool_t (*xdr_argument)(), (*xdr_result)();
	char *(*local)();

	switch (rqstp->rq_proc) {
	case NULLPROC:
		(void) svc_sendreply(transp, xdr_void, (char *)NULL);
		return;

	case RSTATPROC_STATS:
		xdr_argument = xdr_void;
		xdr_result = xdr_statstime;
		local = (char *(*)()) rstatproc_stats_3;
		break;

	case RSTATPROC_HAVEDISK:
		xdr_argument = xdr_void;
		xdr_result = xdr_long;
		local = (char *(*)()) rstatproc_havedisk_3;
		break;

	default:
		svcerr_noproc(transp);
		return;
	}
	bzero((char *)&argument, sizeof(argument));
	if (!svc_getargs(transp, xdr_argument, &argument)) {
		svcerr_decode(transp);
		return;
	}
	result = (*local)(&argument, rqstp);
	if (result != NULL && !svc_sendreply(transp, xdr_result, result)) {
		svcerr_systemerr(transp);
	}
	if (!svc_freeargs(transp, xdr_argument, &argument)) {
		_msgout("unable to free arguments");
		exit(1);
	}
	return;
}

static void
rstatprog_4(rqstp, transp)
	struct svc_req *rqstp;
	register SVCXPRT *transp;
{
	union {
		int fill;
	} argument;
	statsvar *rtn;
	char *result;
	bool_t (*xdr_argument)(), (*xdr_result)();
	char *(*local)();

	switch (rqstp->rq_proc) {
	case NULLPROC:
		(void) svc_sendreply(transp, xdr_void, (char *)NULL);
		return;

	case RSTATPROC_STATS:
		xdr_argument = xdr_void;
		xdr_result = xdr_statsvar;
		local = (char *(*)()) rstatproc_stats_4;
		break;

	case RSTATPROC_HAVEDISK:
		xdr_argument = xdr_void;
		xdr_result = xdr_long;
		local = (char *(*)()) rstatproc_havedisk_4;
		break;

	default:
		svcerr_noproc(transp);
		return;
	}

	bzero((char *)&argument, sizeof(argument));
	if (!svc_getargs(transp, xdr_argument, &argument)) {
		svcerr_decode(transp);
		return;
	}
	result = (*local)(&argument, rqstp);	
#ifdef DEBUG
	rtn = (statsvar *) result;
	putline(rtn->curtime.tv_sec, rtn->boottime, rtn->avenrun);
	if (result == NULL)
	  printf ("*** result is NULL \n");
#endif
 	if (result != NULL && !svc_sendreply(transp, xdr_result, result)) { 

#ifdef DEBUG
	  printf("System error has occured\n");
#endif
		svcerr_systemerr(transp);
	}
	if (!svc_freeargs(transp, xdr_argument, &argument)) {
		_msgout("unable to free arguments");
		exit(1);
	}
	return;
}

/*
 * This is the data structure for storing all the local status.
 */
struct local_stats {
	int cp_time[CPUSTATES];
	int dk_xfer[DK_NDRIVE];
	u_int v_pgpgin;
	u_int v_pgpgout;
	u_int v_pswpin;
	u_int v_pswpout;
	u_int v_intr;
	int if_ipackets;
	int if_ierrors;
	int if_opackets;
	int if_oerrors;
	int if_collisions;
	u_int v_swtch;
	long avenrun[3];
	struct timeval boottime;
	struct timeval curtime;
};
static struct local_stats local_store;

statsswtch *
rstatproc_stats_2()
{
	static statsswtch ret;

#ifdef DEBUG
	_msgout("entering stats_service");
#endif
	sincelastreq = 0;	

	(void) memcpy((char *)&(ret.v_pgpgin), (char *)&(local_store.v_pgpgin),
			 sizeof(ret) - (sizeof(ret.cp_time) + 
					 sizeof(ret.dk_xfer)));
	(void) memcpy((char *)ret.cp_time, 
			 (char *)local_store.cp_time, sizeof(ret.cp_time));
	(void) memcpy((char *)ret.dk_xfer, 
			 (char *)local_store.dk_xfer, sizeof(ret.dk_xfer));
#ifdef DEBUG
	_msgout("returning  stats_service");
#endif
	return(&ret);
}

long *
rstatproc_havedisk_2()
{
	static long have;

	have = havedisk();
	return (&have);
}

statstime *
rstatproc_stats_3()
{
	static statstime ret;

#ifdef DEBUG
	_msgout("entering stats_service");
#endif
	sincelastreq = 0;	

	(void) memcpy((char *)&(ret.v_pgpgin), (char *)&(local_store.v_pgpgin),
			 sizeof(ret) - (sizeof(ret.cp_time) + 
					 sizeof(ret.dk_xfer)));
	(void) memcpy((char *)ret.cp_time, 
			 (char *)local_store.cp_time, sizeof(ret.cp_time));
	(void) memcpy((char *)ret.dk_xfer, 
			 (char *)local_store.dk_xfer, sizeof(ret.dk_xfer));
	return(&ret);
}

long *
rstatproc_havedisk_3()
{
	static long have;

	have = havedisk();
	return (&have);
}

statsvar *
rstatproc_stats_4()
{
	static statsvar ret;

#ifdef DEBUG
	_msgout("entering stats_service - version 4");
#endif
	sincelastreq = 0;
	if (ret.cp_time.cp_time_val ==  NULL) {
		if ((ret.cp_time.cp_time_val =
		 (int *) malloc(sizeof(local_store.cp_time))) == NULL) {

		  _msgout("failed in malloc()");
			exit(1);
		}
		ret.cp_time.cp_time_len = sizeof(local_store.cp_time) / 
		  			  sizeof(int);
		if( (ret.dk_xfer.dk_xfer_val =
		  (int *) malloc(sizeof(local_store.dk_xfer)))  == NULL) {
			_msgout("failed in malloc()");
			exit(1);
		}
	}
	(void) memcpy((char *)&(ret.v_pgpgin), 
		      (char *)&(local_store.v_pgpgin),
		 sizeof(ret) - (sizeof(ret.cp_time) + sizeof(ret.dk_xfer)));
	(void) memcpy((char *)ret.cp_time.cp_time_val,
		(char *) local_store.cp_time, sizeof(local_store.cp_time));
	(void) memcpy((char *)ret.dk_xfer.dk_xfer_val,
		(char *) local_store.dk_xfer, sizeof(local_store.dk_xfer));
	ret.dk_xfer.dk_xfer_len = sizeof(local_store.dk_xfer) / sizeof(int);
#ifdef DEBUG

	printf("v_pgpgin: %d cp_time_val: %d dk_xfer_val: %d \n",
                    ret.v_pgpgin, ret.cp_time.cp_time_val, 
		    ret.dk_xfer.dk_xfer_val);
	putline(ret.curtime.tv_sec, ret.boottime, ret.avenrun);
#endif

	return(&ret);
}

long *
rstatproc_havedisk_4()
{
	static long have;

	have = havedisk();
	return (&have);
}

updatestat()
{
	unsigned long  off, hz;
	int i;
	struct vmmeter sum;
	struct ifnet ifnet;
	struct timeval tm, btm;
	
#ifdef DEBUG
	_msgout("entering updatestat");
#endif

	if ((sincelastreq >= CLOSEDOWN) && _rpcpmstart) {
		do_exit(0);
	}
	sincelastreq++;
	if (lseek(kmem, nl[X_HZ].n_value, 0) == -1) {
		_msgout("can't seek in kmem");
		do_exit(1);
	}
	if (read(kmem, &hz, sizeof hz) != sizeof hz) {
		_msgout("can't read hz from kmem");
		do_exit(1);
	}
	if (lseek(kmem, nl[X_CPTIME].n_value, 0) == -1) {
		_msgout("can't seek in kmem");
		do_exit(1);
	}
 	if (read(kmem, local_store.cp_time, sizeof (local_store.cp_time))
		!= sizeof (local_store.cp_time)) {
		_msgout("can't read cp_time from kmem");
		do_exit(1);
	}
	if (lseek(kmem, nl[X_AVENRUN].n_value, 0) ==-1) {
		_msgout("can't seek in kmem");
		do_exit(1);
	}
#ifdef vax
 	if (read(kmem, avrun, sizeof (avrun)) != sizeof (avrun)) {
		_msgout("can't read avenrun from kmem");
		do_exit(1);
	}
	local_store.avenrun[0] = avrun[0] * FSCALE;
	local_store.avenrun[1] = avrun[1] * FSCALE;
	local_store.avenrun[2] = avrun[2] * FSCALE;
#endif
#ifdef sun
 	if (read(kmem, local_store.avenrun, sizeof (local_store.avenrun))
			!= sizeof (local_store.avenrun)) {
		_msgout("can't read avenrun from kmem");
		do_exit(1);
	}
#endif
	if (lseek(kmem, (unsigned long)nl[X_BOOTTIME].n_value, 0) == -1) {
		_msgout("can't seek in kmem");
		do_exit(1);
	}
 	if (read(kmem, &btm, sizeof (local_store.boottime))
		!= sizeof (local_store.boottime)) {
		_msgout("can't read boottime from kmem");
		do_exit(1);
	}
	local_store.boottime = btm;

#ifdef DEBUG
	{
	char logline[64];

	sprintf(logline, "%d %d %d %d", local_store.cp_time[0],
		local_store.cp_time[1], local_store.cp_time[2], 
		local_store.cp_time[3]);
	_msgout(logline);
	}
#endif

	if (lseek(kmem, (unsigned long)nl[X_SUM].n_value, 0) ==-1) {
		_msgout("can't seek in kmem");
		do_exit(1);
	}
 	if (read(kmem, &sum, sizeof sum) != -1) {
		_msgout("can't read sum from kmem");
		do_exit(1);
	}
	local_store.v_pgpgin = sum.v_pgpgin;
	local_store.v_pgpgout = sum.v_pgpgout;
	local_store.v_pswpin = sum.v_pswpin;
	local_store.v_pswpout = sum.v_pswpout;
	local_store.v_intr = sum.v_intr;
	gettimeofday(&tm, NULL);
	local_store.v_intr -= hz*(tm.tv_sec - btm.tv_sec) +
				hz*(tm.tv_usec - btm.tv_usec)/1000000;
	local_store.v_swtch = sum.v_swtch;

	if (lseek(kmem, (unsigned long)nl[X_DKXFER].n_value, 0) == -1) {
		_msgout("can't seek in kmem");
		do_exit(1);
	}
 	
	if (read(kmem, local_store.dk_xfer, sizeof (local_store.dk_xfer))
			!= sizeof (local_store.dk_xfer)) {
		_msgout("can't read dk_xfer from kmem");
		do_exit(1);
	}

	local_store.if_ipackets = 0;
	local_store.if_opackets = 0;
	local_store.if_ierrors = 0;
	local_store.if_oerrors = 0;
	local_store.if_collisions = 0;
	for (off = firstifnet, i = 0; off && i < numintfs; i++) {
		if (lseek(kmem,  off, 0) == -1) {
			_msgout("can't seek in kmem");
			do_exit(1);
		}
		if (read(kmem, &ifnet, sizeof ifnet) == -1) {
			_msgout("can't read ifnet from kmem");
			do_exit(1);
		}
		local_store.if_ipackets += ifnet.if_ipackets;
		local_store.if_opackets += ifnet.if_opackets;
		local_store.if_ierrors += ifnet.if_ierrors;
		local_store.if_oerrors += ifnet.if_oerrors;
		local_store.if_collisions += ifnet.if_collisions;
		off = (unsigned long) ifnet.if_next;
	}
	gettimeofday(&local_store.curtime, NULL);
	alarm(1);
}

void
setup()
{
	struct ifnet ifnet;
	unsigned long  off;
	
	nlist("/vmunix", nl);
	if (nl[0].n_value == 0) {
		_msgout("Variables missing from namelist");
		do_exit(1);
	}
	if ((kmem = open("/dev/kmem", 0)) < 0) {
		_msgout("can't open kmem");
		do_exit(1);
	}

	off = nl[X_IFNET].n_value;
	if (lseek(kmem, off, 0) == -1) {
		_msgout("can't seek in kmem");
		do_exit(1);
	}
	if (read(kmem, &firstifnet, sizeof( long)) == -1 ) {
		_msgout("can't read firstifnet from kmem");
		do_exit(1);
	}
	numintfs = 0;
	for (off = firstifnet; off;) {
		if (lseek(kmem, off, 0) == -1) {
			_msgout("can't seek in kmem");
			do_exit(1);
		}
		if (read(kmem, &ifnet, sizeof ifnet) != sizeof ifnet) {
			_msgout("can't read ifnet from kmem");
			do_exit(1);
		}
		numintfs++;
		off = (unsigned long) ifnet.if_next;
	}
}

/* 
 * returns true if have a disk
 */
static
havedisk()
{
	int i, cnt;
	long xfer[DK_NDRIVE];

	if (nl[X_DKXFER].n_value == 0) {
		_msgout("Variables missing from namelist");
		do_exit(1);
	}
	if (lseek(kmem, (long)nl[X_DKXFER].n_value, 0) == -1) {
		_msgout("can't seek in kmem");
		do_exit(1);
	}
	if (read(kmem, xfer, sizeof xfer) != sizeof xfer) {
		_msgout("can't read kmem");
		do_exit(1);
	}
	cnt = 0;
	for (i = 0; i < DK_NDRIVE; i++)
		cnt += xfer[i];
	return (cnt != 0);
}

static
do_exit(stat)
	int stat;
{

#ifdef DEBUG
		_msgout("about to closedown");
#endif

#ifndef DEBUG
	if (stat != 0) {
		/*
		 * something went wrong. Sleep awhile and drain our queue
		 * before exiting. This keeps inetd from immediately retrying.
		 */
		int cc, fromlen;
		char buf[RPCSMALLMSGSIZE];
		struct sockaddr from;

		sleep(CLOSEDOWN);
		for (;;) {
			fromlen = sizeof (from);
			cc = recvfrom(0, buf, sizeof (buf), 0, &from, &fromlen);
			if (cc <= 0)
				break;
		}
	}
#endif
	exit(stat);
}

static
_msgout(msg)
	char *msg;
{
#ifdef RPC_SVC_FG
	if (_rpcpmstart)
		syslog(LOG_ERR, msg);
	else
		(void) fprintf(stderr, "%s\n", msg);
#else
	static int initted;

	if (!initted) {
		openlog("rstatd", LOG_PID, LOG_DAEMON);
		initted = 1;
	}
	syslog(LOG_ERR, msg);
#endif
}


#define AVENSIZE (3*sizeof(long))
putline(now, boottime, avenrun)
	time_t now;
	struct timeval boottime;
	long avenrun[];
{
	int uptime, days, hrs, mins, i;
	
	uptime = now - boottime.tv_sec;
	uptime += 30;
	if (uptime < 0)		/* unsynchronized clocks */
		uptime = 0;
	days = uptime / (60*60*24);
	uptime %= (60*60*24);
	hrs = uptime / (60*60);
	uptime %= (60*60);
	mins = uptime / 60;

	printf("  up");
	if (days > 0)
		printf(" %2d day%s", days, days>1?"s,":", ");
	else
		printf("         ");
	if (hrs > 0)
		printf(" %2d:%02d,  ", hrs, mins);
	else
		printf(" %2d min%s", mins, mins>1?"s,":", ");

	/*
	 * Print 1, 5, and 15 minute load averages.
	 * (Found by looking in kernel for avenrun).
	 */
	printf("  load average:");
	for (i = 0; i < (AVENSIZE/sizeof(avenrun[0])); i++) {
		if (i > 0)
			printf(",");
		printf(" %.2f", (double)avenrun[i]/FSCALE);
	}
	printf("\n");
}









